### ISSUE TITLE
Stereo top camera is recorded as one concatenated frame instead of two views

### ISSUE THREAD

**Issue by TheWisp** [TheWisp/lerobot#103]
The OpenArm2 setup's top camera is a ZED stereo unit, and it is recorded as a single frame with both eyes concatenated side by side rather than as two views.

## What is on disk

`observation.images.top` is **2560×720** — two 1280×720 eyes in one frame, an aspect ratio of 3.56:1. Every stereo-shaped camera feature on the rig has exactly that resolution; there is no second format to reconcile.

On fc500t, where the OpenArm2 data actually lives:

```
37 datasets carry a 2560x720 top camera   (27 with episodes in them)
1,121 episodes, 252,835 frames
```

Two caveats on that number, both of which cut it down:

- Many are derivatives of each other, not independent recordings — `0803_20260803_174402`, `…_before_cleanup`, `…_sub64`, `BACKUP_walnut_244ep_used_for_HVLA40k_before_reedit`, several `eval_openarm (copy) (copy)…`. The count of *distinct* captures is far smaller than 37.
- Ten of the 37 have zero episodes.

The duplicates are themselves informative: they exist because editing or filtering a dataset means copying it first by hand (#91), so any conversion has to expect several near-identical copies per real capture and decide whether all of them are worth converting.

## Why the concatenation is the wrong shape

**Most of the frame is redundant.** Two eyes of the same scene, a few centimetres apart, means a model with no stereo-aware structure spends half its input capacity on a near-duplicate of the other half.

**It destroys resolution anisotropically downstream.** Any pipeline that resizes to a square — the 224×224 preparation, among others — takes 2560×720 to 224×224. That is an 11.4× horizontal squeeze against a 3.2× vertical one, so each eye's 1280 columns land in roughly 112 pixels, and the geometry is distorted on top of the loss. Whatever the model learns about shape is learned from a horizontally compressed world.

**Nothing downstream knows it is stereo.** No code in `src/lerobot` mentions ZED or stereo at all, so the frame is treated as one ordinary camera everywhere: sampling, augmentation, encoding, and any crop or flip augmentation will cut across the seam between two different viewpoints.

## Connected: there is no way to exclude an input at training time

Splitting `top` into `top_l` / `top_r` immediately raises the question of training on one eye, or on neither, without re-recording or duplicating the dataset. Today `input_features` is inferred wholesale from the dataset when not set (`# `input_features` can be set to None/null in order to infer those values from the dataset`), and there is no exclusion knob on the training config — so "use every camera in the dataset" is effectively the only option a user can express.

That means the split is not purely additive: doubling the number of top-camera features increases what every policy consumes by default, on datasets that already carry four cameras.

## Open questions worth settling before changing the format

- Whether the right representation is two independent camera features, one feature with a stereo-aware type, or a rectified/depth product computed once at record time — this is the part flagged as pending research, and the ZED SDK's own conventions and how other stacks represent stereo are the obvious places to start.
- Whether existing datasets are converted in place, re-derived into new datasets, or left as they are with the new format applying only going forward — and whether the hand-made duplicates of the same capture are converted alongside the original or dropped.
- What a converted dataset claims about itself, so a model trained on the old concatenated `top` is not silently fed two half-width eyes instead.

**Comment by TheWisp** [comment]
The mechanism for this is now on `feat/hvla-resize-integration`, and the three open questions at the bottom of this issue got settled in the course of building it. Recording the answers here, because they were decided on their merits rather than by reading this issue first — I found it afterwards.

## The open questions, answered

**Representation.** Two independent camera features, `top_l` and `top_r`, named by a suffix on the camera key to match the existing `{camera}_depth` convention for a camera contributing more than one observation channel. Not a stereo-aware type, and not a rectified/depth product at record time.

The ZED SDK route was researched and rejected for now rather than skipped. The normal pipeline is `Camera.open()` → `grab()` → `retrieve_image(VIEW.LEFT)`, which returns a **rectified** left image, with depth available through `retrieve_measure`. It would give rectification (the raw UVC frames are unrectified) and a path to depth. Against that: it requires the proprietary SDK with CUDA, Blackwell needs SDK ≥ 5.1 and CUDA ≥ 12 with an unresolved 5090 report upstream, it would sit resident on the GPU used for training, and — decisively — it cannot convert the episodes already recorded. Worth revisiting if depth is ever wanted; it means re-recording, since rectified frames are different input from what exists.

Both eyes are kept rather than one discarded. Which eye a model consumes is a training-time choice; discarding at record time freezes that decision into data that cannot be re-recorded. It costs nothing: 2560×720 and two 1280×720 frames are the same pixel count.

**Existing datasets.** Re-derived into new datasets, not converted in place and not left alone. Three measurements decided it, against a `stereo_split` flag on the feature that a reader would honour at load time:

- Re-encoding cost is negligible at the size that matters. Measured with the encoder that actually runs (SVT-AV1, CRF 30, the pipeline default): 37.11 dB PSNR against the decoded source, measured at the 224×224 the policy sees. CRF 20 buys 0.03 dB for 44% more storage.
- A flagged dataset stays unmergeable. `merge_into` compares *stored* features, so a 2560-wide `top` would never merge with data recorded natively as `top_l`/`top_r` — and merging is how the current training set was built.
- Converting unifies the codec. The 274-episode merged set was 52 h264 files and 8 AV1; the converted copy is uniformly AV1.

Conversion of the merged 274-episode set: 274 episodes and 47,803 frames preserved, quality-label columns element-wise identical to the source, 9.6 GB → 4.3 GB, about an hour single-threaded. The other datasets counted in this issue are **not** converted; that is the part still outstanding, and the duplicate-derivative problem noted above still applies to deciding which deserve it.

**What a converted dataset claims about itself.** Both keys are renamed. Keeping `top` as the left eye would silently change what an existing name means; renaming makes the mismatch loud, because inference resolves robot cameras by the keys stored in the checkpoint and refuses to start on a missing one. A model trained on the old concatenated `top` fails immediately against a split robot rather than being fed a half-width eye.

## The connected problem

`--cameras` on the S1 trainer answers the exclusion question raised here. Names may be bare or fully qualified, the default is every camera in the dataset so existing recipes are unchanged, and an unknown name is an error rather than a silent no-op. The selection is stored in `image_features` in the checkpoint, which is what inference reads. It is surfaced in the training form as a per-dataset camera picker.

## Confirming the resolution argument

The anisotropy described here is what the work was aimed at, and it measures as stated. Under DINOv2's 14 px patches at 224×224, one patch spans **160×45 px** of a 2560×720 frame; the manipulation target — a ball roughly 60 px across — is therefore *smaller than a single patch horizontally*. One eye halves that to 80×45 and the horizontal squash falls from 3.56× to 1.78×.

Which half is the left eye was measured rather than assumed: cross-correlating the two halves of real rig frames peaks at −72 px and −70 px on two independent frames (correlation 0.51 against 0.19 at zero shift), and since x_R = x_L − d, a negative peak identifies the first half as the left eye. The implied depth, 640 px × 63 mm / 72 px ≈ 560 mm, matches where the tray sits.

## Verified on the rig

A teleoperated recording with the split enabled carries `['left_wrist', 'right_wrist', 'top_l', 'top_r']`, both eyes 720×1280×3, no undivided `top`, one codec. The two channels differ (mean abs difference 0.194 — a duplicated half would satisfy every shape assertion) and cross-correlate at −69 px, matching the offline measurements. A dataset converted offline yields exactly the channel set the robot declares, compared feature-by-feature, which is the property deployment depends on.

One defect only a live recording could have found: a bimanual rig opens top-level cameras through the left arm and strips the `left_` prefix by exact name, so renaming `top` stopped them matching and the first recording came out as `left_top_l`/`left_top_r`. That breaks deployment rather than merely reading oddly, and the bimanual wrapper is the one place the two naming schemes meet.

Relevant commits: `d00269c9a` (contract, camera, robot fan-out, dataset transform, `--cameras`), `e0177ec73` and `f24e8905f` (GUI surfaces), `4a8654da5`, `e63b43388`, `7302de1e2`.

Dropping `discussion`: the format decision is made. Leaving open for the remaining datasets.
